kaldb/src/main/java/com/slack/kaldb/server/ArmeriaService.java

```diff
@@ -38,130 +36,106 @@
-    initializeEventLoop(sb);
-    addCompression(sb);
-    addManagementEndpoints(sb);
-    addTracing(sb);
-    this.server = sb.build();
-  }
-
-  @Override
-  public String toString() {
-    return "ArmeriaService{" + "serviceName='" + serviceName + '\'' + '}';
-  }
-
-  public ArmeriaService(
-      int serverPort,
-      PrometheusMeterRegistry prometheusMeterRegistry,
-      KaldbServiceGrpc.KaldbServiceImplBase searcher,
-      String serviceName) {
-    this.serverPort = serverPort;
-    this.prometheusMeterRegistry = prometheusMeterRegistry;
-    this.serviceName = serviceName;
-
-    ServerBuilder sb = Server.builder();
-    addSearchServices(sb, searcher);
-    initializeEventLoop(sb);
-    addCompression(sb);
-    setTimeout(sb);
-    addManagementEndpoints(sb);
-    addTracing(sb);
-    this.server = sb.build();
-  }
-
-  private void addSearchServices(ServerBuilder sb, KaldbServiceGrpc.KaldbServiceImplBase searcher) {
-    GrpcServiceBuilder searchBuilder =
-        GrpcService.builder()
-            .addService(searcher)
-            .enableUnframedRequests(true)
-            .useBlockingTaskExecutor(true);
-    sb.service(searchBuilder.build());
-    sb.annotatedService(new ElasticsearchApiService(searcher));
   }
 
-  private void initializeEventLoop(ServerBuilder sb) {
-    EventLoopGroup eventLoopGroup = EventLoopGroups.newEventLoopGroup(WORKER_EVENT_LOOP_THREADS);
-    EventLoopGroups.warmUp(eventLoopGroup);
-    sb.workerGroup(eventLoopGroup, true);
-  }
-
-  private void addCompression(ServerBuilder sb) {
-    sb.decorator(EncodingService.builder().newDecorator());
-  }
-
-  private void setTimeout(ServerBuilder sb) {
-    sb.requestTimeout(ARMERIA_TIMEOUT_DURATION);
-    // todo - on timeout explore including a retry-after header via exception handler
-    //  https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503
-    // sb.exceptionHandler()
-  }
-
-  private void addTracing(ServerBuilder sb) {
-    String endpoint = KaldbConfig.get().getTracingConfig().getZipkinEndpoint();
-    SpanHandler spanHandler = new SpanHandler() {};
-
-    if (!endpoint.isBlank()) {
-      LOG.info(String.format("Trace reporting enabled: %s", endpoint));
-      Sender sender = URLConnectionSender.create(endpoint);
-      spanHandler = AsyncZipkinSpanHandler.create(sender);
-    } else {
-      LOG.info("Trace reporting disabled");
+  public static class Builder {
+    private final String serviceName;
+    private final ServerBuilder serverBuilder;
+
+    public Builder(int port, String serviceName, PrometheusMeterRegistry prometheusMeterRegistry) {
+      this.serviceName = serviceName;
+      this.serverBuilder = Server.builder().http(port);
+
+      // initialize event loop
+      EventLoopGroup eventLoopGroup = EventLoopGroups.newEventLoopGroup(WORKER_EVENT_LOOP_THREADS);
+      EventLoopGroups.warmUp(eventLoopGroup);
+      this.serverBuilder.workerGroup(eventLoopGroup, true);
+
+      // timeout
```

Comment:
Refactoring this code is good. But, I think adding these options in the constructor is an anti-pattern. Instead, the following may be a better option:
* Keep the private methods for each of the options as is. 
* Add an additional method like `addDefaultBuildParams` that adds all of these features automatically. 
* Add an additional static methods in ArmeriaService that like build that will take the params you need and build the service want using the builders.

Reply:
For mandatory parameters, as long as there are only a few of them keeping them in the builder constructor is [recommended](https://blogs.oracle.com/javamagazine/post/exploring-joshua-blochs-builder-design-pattern-in-java). In the event there were a significant amount of mandatory params, a step builder would be a better pattern to use here.

I do agree that this ends up with our constructor doing too much though. I went ahead and added back the private methods for initialization, and automatically call them as part of the constructor.

Reply:
Totally not a big deal - but should the `initialize*` private methods also be composable and not return void?

Reply:
As they're private methods that are called from the constructor, I don't think it significantly matters. Our `ServerBuilder` is final, so if we were to use that approach we'd need to create a temporary instance of `ServerBuilder` and pass it to each of the the initializations methods before finally assigning it. That's certainly an option, but I think it ends up being a bit less readable overall.